File: lla2enu.cpp
#include <cmath>
#include <iostream>
#include <iomanip>
#include <stdexcept>

// ==================== 椭球参数 ====================
namespace CGCS2000 {
    constexpr double a = 6378137.0;                 // 长半轴 (m)
    constexpr double f = 1.0 / 298.257222101;       // 扁率
    constexpr double e2 = 2 * f - f * f;            // 第一偏心率平方
    constexpr double deg2rad = M_PI / 180.0;
    constexpr double rad2deg = 180.0 / M_PI;
}

// ==================== 辅助函数 ====================
double normalizeLongitude(double lon_deg) {
    lon_deg = std::fmod(lon_deg, 360.0);
    if (lon_deg >= 180.0) lon_deg -= 360.0;
    if (lon_deg < -180.0) lon_deg += 360.0;
    return lon_deg;
}

// ==================== LLA <-> ECEF ====================
void llaToEcef(double lat_deg, double lon_deg, double alt_m,
               double &X, double &Y, double &Z) {
    using namespace CGCS2000;
    double lat = lat_deg * deg2rad;
    double lon = normalizeLongitude(lon_deg) * deg2rad;
    double sinLat = std::sin(lat), cosLat = std::cos(lat);
    double sinLon = std::sin(lon), cosLon = std::cos(lon);

    double N = a / std::sqrt(1.0 - e2 * sinLat * sinLat);
    double nh = N + alt_m;
    X = nh * cosLat * cosLon;
    Y = nh * cosLat * sinLon;
    Z = (N * (1.0 - e2) + alt_m) * sinLat;
}

void ecefToLla(double X, double Y, double Z,
               double &lat_deg, double &lon_deg, double &alt_m) {
    using namespace CGCS2000;
    double p = std::sqrt(X*X + Y*Y);
    if (p < 1e-12) { // 极轴附近
        lon_deg = 0.0;
        lat_deg = (Z >= 0) ? 90.0 : -90.0;
        alt_m = std::fabs(Z) - a * (1.0 - e2);
        return;
    }

    double lon = std::atan2(Y, X);
    lon_deg = normalizeLongitude(lon * rad2deg);

    double lat = std::atan2(Z, p * (1.0 - e2));
    double prevLat;
    int iter = 0;
    const int MAX_ITER = 30;
    const double CONV = 1e-12;
    do {
        prevLat = lat;
        double sinLat = std::sin(lat), cosLat = std::cos(lat);
        double N = a / std::sqrt(1.0 - e2 * sinLat * sinLat);
        double h = p / cosLat - N;
        lat = std::atan2(Z, p * (1.0 - e2 * N / (N + h)));
        iter++;
    } while (std::fabs(lat - prevLat) > CONV && iter < MAX_ITER);

    if (iter >= MAX_ITER)
        throw std::runtime_error("ECEF->LLA 迭代未收敛");

    double sinLat = std::sin(lat), cosLat = std::cos(lat);
    double N = a / std::sqrt(1.0 - e2 * sinLat * sinLat);
    if (std::fabs(cosLat) < 1e-12)
        alt_m = Z / sinLat - N * (1.0 - e2);
    else
        alt_m = p / cosLat - N;

    lat_deg = lat * rad2deg;
}

// ==================== LLA <-> ENU 直接接口 ====================
/**
 * 将目标点的大地坐标 (lat, lon, alt) 转换为相对于参考点 (ref_lat, ref_lon, ref_alt) 的 ENU 坐标
 */
void llaToEnu(double lat_deg, double lon_deg, double alt_m,
              double ref_lat_deg, double ref_lon_deg, double ref_alt_m,
              double &E, double &N, double &U) {
    // 1. 两点转ECEF
    double X, Y, Z, X0, Y0, Z0;
    llaToEcef(lat_deg, lon_deg, alt_m, X, Y, Z);
    llaToEcef(ref_lat_deg, ref_lon_deg, ref_alt_m, X0, Y0, Z0);

    // 2. 相对向量
    double dx = X - X0;
    double dy = Y - Y0;
    double dz = Z - Z0;

    // 3. 旋转矩阵（基于参考点）
    using namespace CGCS2000;
    double lat = ref_lat_deg * deg2rad;
    double lon = ref_lon_deg * deg2rad;
    double sinLat = std::sin(lat), cosLat = std::cos(lat);
    double sinLon = std::sin(lon), cosLon = std::cos(lon);

    E = -sinLon * dx + cosLon * dy;
    N = -sinLat * cosLon * dx - sinLat * sinLon * dy + cosLat * dz;
    U =  cosLat * cosLon * dx + cosLat * sinLon * dy + sinLat * dz;
}

/**
 * 将 ENU 坐标转换为目标点的大地坐标 (lat, lon, alt)
 */
void enuToLla(double E, double N, double U,
              double ref_lat_deg, double ref_lon_deg, double ref_alt_m,
              double &lat_deg, double &lon_deg, double &alt_m) {
    // 1. 参考点转ECEF
    double X0, Y0, Z0;
    llaToEcef(ref_lat_deg, ref_lon_deg, ref_alt_m, X0, Y0, Z0);

    // 2. 逆旋转（R^T）
    using namespace CGCS2000;
    double lat = ref_lat_deg * deg2rad;
    double lon = ref_lon_deg * deg2rad;
    double sinLat = std::sin(lat), cosLat = std::cos(lat);
    double sinLon = std::sin(lon), cosLon = std::cos(lon);

    double dx = -sinLon * E - sinLat * cosLon * N + cosLat * cosLon * U;
    double dy =  cosLon * E - sinLat * sinLon * N + cosLat * sinLon * U;
    double dz =                        cosLat * N + sinLat * U;

    // 3. 目标点ECEF
    double X = X0 + dx;
    double Y = Y0 + dy;
    double Z = Z0 + dz;

    // 4. ECEF -> LLA
    ecefToLla(X, Y, Z, lat_deg, lon_deg, alt_m);
}

// ==================== 测试 ====================
int main() {
    // 参考点：北京天安门 (39.9042°N, 116.4074°E, 椭球高50m)
    double ref_lat = 39.9042, ref_lon = 116.4074, ref_alt = 50.0;

    // 目标点：参考点正东100m，正北200m，高10m（ENU坐标）
    double E_test = 100.0, N_test = 200.0, U_test = 10.0;

    // ENU -> LLA
    double lat, lon, alt;
    enuToLla(E_test, N_test, U_test, ref_lat, ref_lon, ref_alt, lat, lon, alt);

    // LLA -> ENU 恢复
    double E_rec, N_rec, U_rec;
    llaToEnu(lat, lon, alt, ref_lat, ref_lon, ref_alt, E_rec, N_rec, U_rec);

    std::cout << std::setprecision(12) << std::fixed;
    std::cout << "参考点: lat=" << ref_lat << "°, lon=" << ref_lon << "°, alt=" << ref_alt << " m\n";
    std::cout << "原始ENU: E=" << E_test << " m, N=" << N_test << " m, U=" << U_test << " m\n";
    std::cout << "\nENU -> LLA:\n";
    std::cout << "Lat = " << lat << " °\n";
    std::cout << "Lon = " << lon << " °\n";
    std::cout << "Alt = " << alt << " m\n";
    std::cout << "\nLLA -> ENU (恢复):\n";
    std::cout << "E = " << E_rec << " m\n";
    std::cout << "N = " << N_rec << " m\n";
    std::cout << "U = " << U_rec << " m\n";

    std::cout << "\n误差 (恢复后与原值比较):\n";
    std::cout << "ΔE = " << (E_rec - E_test) * 1000 << " mm\n";
    std::cout << "ΔN = " << (N_rec - N_test) * 1000 << " mm\n";
    std::cout << "ΔU = " << (U_rec - U_test) * 1000 << " mm\n";

    return 0;
}

File: lla2enu.py
import math

# ==================== 椭球参数 ====================
class CGCS2000:
    a = 6378137.0
    f = 1.0 / 298.257222101
    e2 = 2*f - f*f
    deg2rad = math.pi / 180.0
    rad2deg = 180.0 / math.pi

    @staticmethod
    def normalize_longitude(lon_deg):
        lon_deg = lon_deg % 360.0
        if lon_deg >= 180.0:
            lon_deg -= 360.0
        if lon_deg < -180.0:
            lon_deg += 360.0
        return lon_deg

# ==================== LLA <-> ECEF ====================
def lla_to_ecef(lat_deg, lon_deg, alt_m):
    a = CGCS2000.a
    e2 = CGCS2000.e2
    lat = math.radians(lat_deg)
    lon = math.radians(CGCS2000.normalize_longitude(lon_deg))
    sin_lat = math.sin(lat)
    cos_lat = math.cos(lat)
    sin_lon = math.sin(lon)
    cos_lon = math.cos(lon)
    N = a / math.sqrt(1.0 - e2 * sin_lat * sin_lat)
    nh = N + alt_m
    X = nh * cos_lat * cos_lon
    Y = nh * cos_lat * sin_lon
    Z = (N * (1.0 - e2) + alt_m) * sin_lat
    return X, Y, Z

def ecef_to_lla(X, Y, Z):
    a = CGCS2000.a
    e2 = CGCS2000.e2
    p = math.hypot(X, Y)
    if p < 1e-12:  # 极轴附近
        lon_deg = 0.0
        lat_deg = 90.0 if Z >= 0 else -90.0
        alt_m = abs(Z) - a * (1.0 - e2)
        return lat_deg, lon_deg, alt_m

    lon = math.atan2(Y, X)
    lon_deg = math.degrees(lon)
    lon_deg = CGCS2000.normalize_longitude(lon_deg)

    lat = math.atan2(Z, p * (1.0 - e2))
    for _ in range(30):
        prev_lat = lat
        sin_lat = math.sin(lat)
        cos_lat = math.cos(lat)
        N = a / math.sqrt(1.0 - e2 * sin_lat * sin_lat)
        h = p / cos_lat - N
        lat = math.atan2(Z, p * (1.0 - e2 * N / (N + h)))
        if abs(lat - prev_lat) < 1e-12:
            break
    else:
        raise RuntimeError("ECEF->LLA 迭代未收敛")

    sin_lat = math.sin(lat)
    cos_lat = math.cos(lat)
    N = a / math.sqrt(1.0 - e2 * sin_lat * sin_lat)
    if abs(cos_lat) < 1e-12:
        alt_m = Z / sin_lat - N * (1.0 - e2)
    else:
        alt_m = p / cos_lat - N

    lat_deg = math.degrees(lat)
    return lat_deg, lon_deg, alt_m

# ==================== LLA <-> ENU 直接接口 ====================
def lla_to_enu(lat_deg, lon_deg, alt_m,
               ref_lat_deg, ref_lon_deg, ref_alt_m):
    """
    将目标点大地坐标转换为相对于参考点的 ENU 坐标
    返回 (E, N, U)
    """
    X, Y, Z = lla_to_ecef(lat_deg, lon_deg, alt_m)
    X0, Y0, Z0 = lla_to_ecef(ref_lat_deg, ref_lon_deg, ref_alt_m)

    dx = X - X0
    dy = Y - Y0
    dz = Z - Z0

    lat = math.radians(ref_lat_deg)
    lon = math.radians(ref_lon_deg)
    sin_lat = math.sin(lat)
    cos_lat = math.cos(lat)
    sin_lon = math.sin(lon)
    cos_lon = math.cos(lon)

    E = -sin_lon * dx + cos_lon * dy
    N = -sin_lat * cos_lon * dx - sin_lat * sin_lon * dy + cos_lat * dz
    U =  cos_lat * cos_lon * dx + cos_lat * sin_lon * dy + sin_lat * dz
    return E, N, U

def enu_to_lla(E, N, U,
               ref_lat_deg, ref_lon_deg, ref_alt_m):
    """
    将 ENU 坐标转换为目标点的大地坐标
    返回 (lat_deg, lon_deg, alt_m)
    """
    X0, Y0, Z0 = lla_to_ecef(ref_lat_deg, ref_lon_deg, ref_alt_m)

    lat = math.radians(ref_lat_deg)
    lon = math.radians(ref_lon_deg)
    sin_lat = math.sin(lat)
    cos_lat = math.cos(lat)
    sin_lon = math.sin(lon)
    cos_lon = math.cos(lon)

    dx = -sin_lon * E - sin_lat * cos_lon * N + cos_lat * cos_lon * U
    dy =  cos_lon * E - sin_lat * sin_lon * N + cos_lat * sin_lon * U
    dz =                        cos_lat * N + sin_lat * U

    X = X0 + dx
    Y = Y0 + dy
    Z = Z0 + dz

    return ecef_to_lla(X, Y, Z)

# ==================== 测试 ====================
if __name__ == "__main__":
    ref_lat, ref_lon, ref_alt = 39.9042, 116.4074, 50.0
    E_test, N_test, U_test = 100.0, 200.0, 10.0

    lat, lon, alt = enu_to_lla(E_test, N_test, U_test, ref_lat, ref_lon, ref_alt)
    E_rec, N_rec, U_rec = lla_to_enu(lat, lon, alt, ref_lat, ref_lon, ref_alt)

    print(f"参考点: lat={ref_lat}°, lon={ref_lon}°, alt={ref_alt}m")
    print(f"原始ENU: E={E_test}m, N={N_test}m, U={U_test}m")
    print("\nENU -> LLA:")
    print(f"Lat = {lat:.12f} °")
    print(f"Lon = {lon:.12f} °")
    print(f"Alt = {alt:.6f} m")
    print("\nLLA -> ENU (恢复):")
    print(f"E = {E_rec:.6f} m")
    print(f"N = {N_rec:.6f} m")
    print(f"U = {U_rec:.6f} m")
    print("\n误差 (恢复后与原值比较):")
    print(f"ΔE = {(E_rec - E_test)*1000:.3f} mm")
    print(f"ΔN = {(N_rec - N_test)*1000:.3f} mm")
    print(f"ΔU = {(U_rec - U_test)*1000:.3f} mm")